[Metabot] Add streaming support

sloansparger opened the pull request:
Part of [BOT-115](https://linear.app/metabase/issue/BOT-115/frontend-ai-sdk-streaming-implementation)

### Description

This PR adds streaming support to the frontend of metabot. Currently it has to be enabled by double-clicking the icon next to the prompt input like so:
https://github.com/user-attachments/assets/30fd40fd-18fa-4bde-a840-3d97eed93698
You conversation state will be cleared and once metabot is yellow streaming support is on.


### How to verify

There's no specific script to encourage folks to go down, but here's a broad few ideas to explore in making sure it's working:
- conversation state + history builds up over time like you'd expect
- normal chat conversation behave how they did before
- you're able to see metabot's in progress work
- failed requests work like you'd expect

### Demo

https://www.loom.com/share/b837bd234bae432a9d4b0cd99d00cfe6?sid=a02edad6-ac40-41fa-8a96-9293058b47f2

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

sloansparger committed:
add BE code

sloansparger committed:
port over most of code from branch metabot/streaming w/o tests + making changes for supporting both normal requests and streamed ones

sloansparger committed:
improve test coverage, bug fixes, add support for analyze chart, error handling changes

sloansparger committed:
small bit of clean up + add unit tests for streaming

Somtom commented on enterprise/backend/src/metabase_enterprise/metabot_v3/api.clj:
```diff
@@ -59,6 +59,36 @@
+  "Send a chat message to the LLM via the AI Proxy."
+  [_route-params
+   _query-params
+   ; TODO: conversation_id is unused, should we should send it back as a header?
```
**question:** What do you mean with "unused"? The conversation id should somehow reach ai-service so that we can use it for tracking purposes

Somtom commented on enterprise/frontend/src/metabase-enterprise/metabot/components/MetabotChat/MetabotChat.tsx:
```diff
@@ -282,3 +288,49 @@ const Message = ({
+      t`Lot's to consider`,
+      t`Working through the details...`,
+    ],
+    construct_notebook_query: [t`Creating a query`, t`Contructing a question`],
+    analyze_data: [t`Analyzing the data`, t`Exploring your data`],
+    analyze_chart: [t`Inspecting the visualization`, t`Looking at the data`],
+    list_available_fields: undefined, // tool executes near instantly
```
**Suggestion(not-blocking):** Do we maybe want to control this from ai-service side? The names of the tools could change easily there without Metabase knowing it (i.e. if we optimize them for the LLM). I was thinking about this in the past and wondering whether we wanted some kind of user message like those here on the ai-service side that we define on the tool level.
This way, if the tool names change, we don't have an issue with the mapping.

Not blocking here - we can discuss / iterate on this separately.

retro replied:
We probably want to validate if we know how to handle a tool name and show a generic message otherwise. That way we can still apply i18n to the messages

sloansparger replied:
That's how it's designed now, so if that's what we want it's good as is.

Long term, will we want strings translated in the ai-service? If so, then I think Thomas' suggestion still stands if we introduce a ttag / trs like solution there.

sloansparger committed:
fix some subtle bugs around request cancellation and stream processing

Somtom commented on enterprise/frontend/src/metabase-enterprise/metabot/state/reducer.ts:
```diff
@@ -17,29 +17,40 @@ export type MetabotChatMessage =
+        state.history.push({ role: "user", content: action.payload });
+      }
     },
     addAgentMessage: (
```
**issue (not-blocking):** As soon as we start streaming text chunks of a single message, this would break and push the text chunks as individual messages.

For example, consider the following response (I prototyped parts of the ai-service side and this is a request we could expect):
```
0:"I"
0:" see"
0:" we"
0:" have"
0:" an"
0:" \""
0:"Orders"
0:"\""
0:" table"
0:" available"
0:","
0:" but"
0:" I"
0:" need"
0:" to"
0:" confirm"
0:" if"
0:" there"
0:" is"
0:" a"
0:" field"
0:" that"
0:" identifies"
0:" the"
0:" location"
0:" or"
0:" region"
0:" of"
0:" the"
0:" orders"
0:","
0:" specifically"
0:" something"
0:" that"
0:" can"
0:" indicate"
0:" \""
0:"east"
0:" coast"
0:".\""
0:" Let"
0:" me"
0:" first"
0:" check"
0:" the"
0:" fields"
0:" in"
0:" the"
0:" Orders"
0:" table"
0:" to"
0:" see"
0:" if"
0:" there's"
0:" a"
0:" relevant"
0:" location"
0:" or"
0:" region"
0:" field"
0:"."
9:{"toolCallId":"call_ZqXPhdDXxUIc6Z6PRrSsOsm9","toolName":"list_available_fields","args":"{\"tables_ids\":[\"card__113\"],\"metric_ids\":[]}"}
a:{"toolCallId":"call_ZqXPhdDXxUIc6Z6PRrSsOsm9","result":"\n\n\n<tables>\n\n\n<table id=\"card__113\", name=\"Orders\">\n\n### Description\nNone\n\n### Fields\n\nThe following fields are available in this table:\n\n| Field Name | Field ID | Type | Description |\n| ---------- | -------- | ---- | ----------- |\n| ID | c113/0 | number | This is a unique ID for the product. It is also called the \u201cInvoice number\u201d or \u201cConfirmation number\u201d in customer facing emails and screens. |\n| User ID | c113/1 | number | The id of the user who made this order. Note that in some cases where an order was created on behalf of a customer who phoned the order in, this might be the employee who handled the request. |\n| Product ID | c113/2 | number | The product ID. This is an internal identifier for the product, NOT the SKU. |\n| Subtotal | c113/3 | number | The raw, pre-tax cost of the order. Note that this might be different in the future from the product price due to promotions, credits, etc. |\n| Tax | c113/4 | number | This is the amount of local and federal taxes that are collected on the purchase. Note that other governmental fees on some products are not included here, but instead are accounted for in the subtotal. |\n| Total | c113/5 | number | The total billed amount. |\n| Discount | c113/6 | number | Discount amount. |\n| Created At | c113/7 | datetime | The date and time an order was submitted. |\n| Quantity | c113/8 | number | Number of products bought. |\n| User ID | c113/9 | string | The name of the user who owns an account |\n| Product ID | c113/10 | string | The name of the product as it should be displayed to customers. |\n\n</table>\n\n</tables>\n"}
0:"The"
0:" Orders"
0:" table"
0:" has"
0:" fields"
0:" for"
0:" order"
0:" details"
0:" like"
0:" ID"
0:","
0:" user"
0:","
0:" product"
0:","
0:" amounts"
0:","
```

You see thee response can also include tool calls inbetween.

There is also the issue with the tool calls that might happen while streaming. I guess for those we might want to keep some indicator in the text that here a tool call happened.

I was trying to quickly hack something on the FE to visualize it
<img width="525" alt="image" src="https://github.com/user-attachments/assets/196e70a3-98a4-408f-a8d5-f7c306208e5a" />


This is not blocking at all but something we should consider moving forward with the streaming (as soon as we have these text chunks on ai-service side supported)

sloansparger replied:
Great point. We just need to make sure this is correct before making streaming the default. Going to correct this in a follow up PR.

Somtom replied:
Follow-up PR is entirely fine here. Should not block this one. 👍

Somtom reviewed (approved):
I tested things locally and it worked very well so far.

I was exploring how we could better stream the text chunks from ai-service side and this led to some potential issue for the future (but not blocking this PR). See other comment.

heypoom reviewed (approved):
Changes to `MetabotQuestion.tsx` looks good 🙏🏻

retro committed:
Remove debugging logs

sloansparger committed:
fix be lints and things

camsaul commented on enterprise/backend/src/metabase_enterprise/metabot_v3/client.clj:
```diff
@@ -129,6 +134,57 @@
+          (loop []
+            (when-let [line (.readLine response-lines)]
+              (.write os (.getBytes (str line "\n") "UTF-8"))
+              (.flush os)
```
Are you sure you want to be flushing on every single line? Let the `streaming-response` code decide when to buffer and when to flush.

camsaul commented on enterprise/backend/src/metabase_enterprise/metabot_v3/client.clj:
```diff
@@ -129,6 +134,57 @@
+              (.write os (.getBytes (str line "\n") "UTF-8"))
+              (.flush os)
+                                   ;; Give time for the line to flush before continuing
+              (Thread/sleep 10)
```
Can't you just use something like `org.apache.commons.io.IOUtils/copy` for this??

camsaul reviewed (changes requested):
- **Huge** no-no to open up a `java.io.Reader` and not close it (use `with-open` here)
- Streaming code seems 100x more complicated than it needs to be and the whole sleep 10ms seems like a massive code smell

retro committed:
Improve ai service response streaming code

- Wrap initialization of response-lines reader with with-open to automatically close it after the response is done
- Remove Thread/sleep code

retro committed:
Update modules config to register usage of the server module

retro committed:
Remove request body destructure in the agent-streaming endpoint

retro committed:
Fix reflection error

retro committed:
Expand the docstring for the streaming-request function

sloansparger committed:
sort namespaces
